examples/custom-congestion-controller/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Example implementation of a custom congestion controller algorithm.
+///
+/// This example serves only to illustrate the integration points for incorporating a custom
+/// congestion controller into s2n-quic, and not as an actual congestion controller implementation.
+///
+/// NOTE: The `CongestionController` trait is considered unstable and may be subject to change
+///       in a future release.
+pub mod custom_congestion_controller {
+    use s2n_quic::provider::{
+        congestion_controller,
+        congestion_controller::{
+            CongestionController, Publisher, RandomGenerator, RttEstimator, Timestamp,
+        },
+    };
+
+    /// Define a congestion controller containing any state you wish to track.
+    /// For this example, we track the size of the congestion window in bytes and
+    /// the number of bytes in flight.
+    #[derive(Debug, Clone)]
+    pub struct MyCongestionController {
+        congestion_window: u32,
+        bytes_in_flight: u32,
+    }
+
+    /// The following is a simple implementation of the `CongestionController` trait
+    /// that increases the congestion window by the number of bytes acknowledged and
+    /// decreases the congestion window by half when packets are lost.
+    #[allow(unused)]
+    impl CongestionController for MyCongestionController {
+        // A custom `PacketInfo` type may optionally be defined to include additional per-packet
+        // state. This state will be stored for each in flight packet, and returned to the
+        // `on_ack` and `on_packet_lost` methods.
+        type PacketInfo = ();
```

Comment:
I forget did we have any tests that enforces that this was < than some size. can that test break if users provide a large type here?

Reply:
We only check that the `SentPacketInfo` with no additional `PacketInfo` isn't too big: https://github.com/aws/s2n-quic/blob/1cb037186ef49b594bc5f5bc4f99e934b614cd46/quic/s2n-quic-core/src/recovery/sent_packets.rs#L98